### ISSUE TITLE
Docker fails on both ElasticSearch, Redis and RabbitMQ

### ISSUE THREAD

**Issue by Asitis** [markshust/docker-magento#629]
I've been using this docker setup for a few years now. But I had some permission issues,  and after I ran `bin/update`, I couldn't restart the containers due to all kinds of issues, so I decided to remove everything and start over with a clean slate on Ubuntu.
I tried the one-line install and the manual install, but neither seem to function anymore from `bin/setup magento2.test` forward. I already checked https://github.com/markshust/docker-magento/issues/442.

**What is happening**
I can download and start the setup just fine; all services are started normally without issue.
First I couldn't connect to Elasticsearch. If I remove `--silent` from the setup file, it throws;
`Failed to connect to elasticsearch port 9200: Connection refused`

If I comment out Elasticsearch, RabbitMQ and Redis throw this issue;
```Waiting for connection to RabbitMQ...
..........Could not connect to the Amqp Server.
SQLSTATE[HY000] [2002] Connection refused
Invalid Redis configuration. Could not connect to Redis server.
Invalid Redis configuration. Could not connect to Redis server.

In InstallCommand.php line 274:
                               
  Parameter validation failed  

setup:install [--backend-frontname BACKEND-FRONTNAME] [--enable-debug-logging ENABLE-DEBUG-LOGGING] [--enable-syslog-logging ENABLE-SYSLOG-LOGGING] [--remote-storage-driver REMOTE-STORAGE-DRIVER] [--remote-storage-prefix REMOTE-STORAGE-PREFIX] [--remote-storage-bucket REMOTE-STORAGE-BUCKET] [--remote-storage-region REMOTE-STORAGE-REGION] [--remote-storage-key REMOTE-STORAGE-KEY] [--remote-storage-secret REMOTE-STORAGE-SECRET] [--consumers-wait-for-messages CONSUMERS-WAIT-FOR-MESSAGES] [--amqp-host AMQP-HOST] [--amqp-port AMQP-PORT] [--amqp-user AMQP-USER] [--amqp-password AMQP-PASSWORD] [--amqp-virtualhost AMQP-VIRTUALHOST] [--amqp-ssl AMQP-SSL] [--amqp-ssl-options AMQP-SSL-OPTIONS] [--key KEY] [--db-host DB-HOST] [--db-name DB-NAME] [--db-user DB-USER] [--db-engine DB-ENGINE] [--db-password DB-PASSWORD] [--db-prefix DB-PREFIX] [--db-model DB-MODEL] [--db-init-statements DB-INIT-STATEMENTS] [-s|--skip-db-validation] [--http-cache-hosts HTTP-CACHE-HOSTS] [--db-ssl-key DB-SSL-KEY] [--db-ssl-cert DB-SSL-CERT] [--db-ssl-ca DB-SSL-CA] [--db-ssl-verify] [--session-save SESSION-SAVE] [--session-save-redis-host SESSION-SAVE-REDIS-HOST] [--session-save-redis-port SESSION-SAVE-REDIS-PORT] [--session-save-redis-password SESSION-SAVE-REDIS-PASSWORD] [--session-save-redis-timeout SESSION-SAVE-REDIS-TIMEOUT] [--session-save-redis-persistent-id SESSION-SAVE-REDIS-PERSISTENT-ID] [--session-save-redis-db SESSION-SAVE-REDIS-DB] [--session-save-redis-compression-threshold SESSION-SAVE-REDIS-COMPRESSION-THRESHOLD] [--session-save-redis-compression-lib SESSION-SAVE-REDIS-COMPRESSION-LIB] [--session-save-redis-log-level SESSION-SAVE-REDIS-LOG-LEVEL] [--session-save-redis-max-concurrency SESSION-SAVE-REDIS-MAX-CONCURRENCY] [--session-save-redis-break-after-frontend SESSION-SAVE-REDIS-BREAK-AFTER-FRONTEND] [--session-save-redis-break-after-adminhtml SESSION-SAVE-REDIS-BREAK-AFTER-ADMINHTML] [--session-save-redis-first-lifetime SESSION-SAVE-REDIS-FIRST-LIFETIME] [--session-save-redis-bot-first-lifetime SESSION-SAVE-REDIS-BOT-FIRST-LIFETIME] [--session-save-redis-bot-lifetime SESSION-SAVE-REDIS-BOT-LIFETIME] [--session-save-redis-disable-locking SESSION-SAVE-REDIS-DISABLE-LOCKING] [--session-save-redis-min-lifetime SESSION-SAVE-REDIS-MIN-LIFETIME] [--session-save-redis-max-lifetime SESSION-SAVE-REDIS-MAX-LIFETIME] [--session-save-redis-sentinel-master SESSION-SAVE-REDIS-SENTINEL-MASTER] [--session-save-redis-sentinel-servers SESSION-SAVE-REDIS-SENTINEL-SERVERS] [--session-save-redis-sentinel-verify-master SESSION-SAVE-REDIS-SENTINEL-VERIFY-MASTER] [--session-save-redis-sentinel-connect-retries SESSION-SAVE-REDIS-SENTINEL-CONNECT-RETRIES] [--cache-backend CACHE-BACKEND] [--cache-backend-redis-server CACHE-BACKEND-REDIS-SERVER] [--cache-backend-redis-db CACHE-BACKEND-REDIS-DB] [--cache-backend-redis-port CACHE-BACKEND-REDIS-PORT] [--cache-backend-redis-password CACHE-BACKEND-REDIS-PASSWORD] [--cache-backend-redis-compress-data CACHE-BACKEND-REDIS-COMPRESS-DATA] [--cache-backend-redis-compression-lib CACHE-BACKEND-REDIS-COMPRESSION-LIB] [--cache-id-prefix CACHE-ID-PREFIX] [--allow-parallel-generation] [--page-cache PAGE-CACHE] [--page-cache-redis-server PAGE-CACHE-REDIS-SERVER] [--page-cache-redis-db PAGE-CACHE-REDIS-DB] [--page-cache-redis-port PAGE-CACHE-REDIS-PORT] [--page-cache-redis-password PAGE-CACHE-REDIS-PASSWORD] [--page-cache-redis-compress-data PAGE-CACHE-REDIS-COMPRESS-DATA] [--page-cache-redis-compression-lib PAGE-CACHE-REDIS-COMPRESSION-LIB] [--page-cache-id-prefix PAGE-CACHE-ID-PREFIX] [--lock-provider LOCK-PROVIDER] [--lock-db-prefix LOCK-DB-PREFIX] [--lock-zookeeper-host LOCK-ZOOKEEPER-HOST] [--lock-zookeeper-path LOCK-ZOOKEEPER-PATH] [--lock-file-path LOCK-FILE-PATH] [--document-root-is-pub DOCUMENT-ROOT-IS-PUB] [--base-url BASE-URL] [--language LANGUAGE] [--timezone TIMEZONE] [--currency CURRENCY] [--use-rewrites USE-REWRITES] [--use-secure USE-SECURE] [--base-url-secure BASE-URL-SECURE] [--use-secure-admin USE-SECURE-ADMIN] [--admin-use-security-key ADMIN-USE-SECURITY-KEY] [--admin-user [ADMIN-USER]] [--admin-password [ADMIN-PASSWORD]] [--admin-email [ADMIN-EMAIL]] [--admin-firstname [ADMIN-FIRSTNAME]] [--admin-lastname [ADMIN-LASTNAME]] [--search-engine SEARCH-ENGINE] [--elasticsearch-host ELASTICSEARCH-HOST] [--elasticsearch-port ELASTICSEARCH-PORT] [--elasticsearch-enable-auth ELASTICSEARCH-ENABLE-AUTH] [--elasticsearch-username ELASTICSEARCH-USERNAME] [--elasticsearch-password ELASTICSEARCH-PASSWORD] [--elasticsearch-index-prefix ELASTICSEARCH-INDEX-PREFIX] [--elasticsearch-timeout ELASTICSEARCH-TIMEOUT] [--cleanup-database] [--sales-order-increment-prefix SALES-ORDER-INCREMENT-PREFIX] [--use-sample-data] [--enable-modules [ENABLE-MODULES]] [--disable-modules [DISABLE-MODULES]] [--convert-old-scripts [CONVERT-OLD-SCRIPTS]] [-i|--interactive] [--safe-mode [SAFE-MODE]] [--data-restore [DATA-RESTORE]] [--dry-run [DRY-RUN]] [--magento-init-params MAGENTO-INIT-PARAMS]
```
**What I already did**
I did a complete purge of docker, docker-compose, this entire install, rebooted, did a reinstall of everything - fails at the same point each time.

**My system**
I'm running Kubuntu 21.10, Kernel 5.13.0, 16GB RAM with a i7-1065G7 CPU, so it's no shortage of muscle on my machine I believe. I did do all the extra Linux setup to resolve the hostname, set the vm.max_map_count=262144 etc.
On top of this, I had the setup running (albeit with permission issues) before on this same machine for a long time.

**So what is happening?**
It looks like there are internal connection issues in the container, neither Elasticsearch, RabbitMQ or Redis can connect properly. I'm too novice with Docker to properly figure out how to fix this. I have been Googling around, but I couldn't find a proper fix. The strange thing is, I had this Docker running before on this same machine. Am I correct in thinking there is something wrong _in_ the container, and not my local machine?

**Comment by markshust** [comment]
If all services start normally without issue, then the issue is most likely related to a configuration which is incorrect.

Have you seen the Linux-related spot in the README? You must follow these updates in order to continue. The oneliner won't work with Linux.

https://github.com/markshust/docker-magento#linux

**Comment by Asitis** [comment]
Hi Mark, thanks for your response. I followed all the Linux steps, verified Docker works, install docker-compose, everything from the Docker Docs and yours, but still the process keeps failing on 'Waiting for connection to Elasticsearch' every time. I've tried again just now, to be sure, but it's still not completing the manual install.

To reiterate I;
- Uninstalled this project, removed all docker instances
- Reinstalled Docker and Docker Compose from scratch, with post-install fixes for Docker
- Created new directory to install this project in
- Ran `curl -s https://raw.githubusercontent.com/markshust/docker-magento/master/lib/template | bash` to get the template
- Ran `bin/download 2.4.3-p1` to download Magento
- Added `host.docker.internal:172.17.0.1` to the app.extra_hosts parameter of docker-compose.yml (with correct IP)
- Added `172.17.0.1 host.docker.internal` to hostfile
- Set `vm.max_map_count=262144` on my system
- Ran `bin/setup magento.test`
- It fails on 'Waiting for connection to Elasticsearch' 

Plus, as I said in the other topic: Using docker stats I see that elasticsearch has 15GB to play around, but it stops after memory usage creeps up to 1.27GB.

Any ideas?

**Comment by markshust** [comment]
@Asitis just to confirm, does this line:

`docker network inspect bridge --format='{{(index .IPAM.Config 0).Gateway}}'`

...output 172.17.0.1?

**Comment by Asitis** [comment]
Yep! Sorry if that wasn't clear!

**Comment by Navari** [comment]
i have the same problem. can anyone solve this problem ?

**Comment by Asitis** [comment]
Thats good to hear I'm not the only one, but could you maybe elaborate on your system and setup, things you did and tried? This way hopefully Mark can zero in on the issue.

**Comment by Navari** [comment]
> Thats good to hear I'm not the only one, but could you maybe elaborate on your system and setup, things you did and tried? This way hopefully Mark can zero in on the issue.

i found the answer, just increase docker RAM

**Comment by Asitis** [comment]
Great that that works for you, sadly, that isn't the issue on my part;
![image](https://user-images.githubusercontent.com/9464641/152498811-202077fa-d799-435a-a72c-8f68231cc43f.png)

**Comment by lalittmohan** [comment]
> Hi Mark, thanks for your response. I followed all the Linux steps, verified Docker works, install docker-compose, everything from the Docker Docs and yours, but still the process keeps failing on 'Waiting for connection to Elasticsearch' every time. I've tried again just now, to be sure, but it's still not completing the manual install.
> 
> To reiterate I;
> 
>     * Uninstalled this project, removed all docker instances
> 
>     * Reinstalled Docker and Docker Compose from scratch, with post-install fixes for Docker
> 
>     * Created new directory to install this project in
> 
>     * Ran `curl -s https://raw.githubusercontent.com/markshust/docker-magento/master/lib/template | bash` to get the template
> 
>     * Ran `bin/download 2.4.3-p1` to download Magento
> 
>     * Added `host.docker.internal:172.17.0.1` to the app.extra_hosts parameter of docker-compose.yml (with correct IP)
> 
>     * Added `172.17.0.1 host.docker.internal` to hostfile
> 
>     * Set `vm.max_map_count=262144` on my system
> 
>     * Ran `bin/setup magento.test`
> 
>     * It fails on 'Waiting for connection to Elasticsearch'
> 
> 
> Plus, as I said in the other topic: Using docker stats I see that elasticsearch has 15GB to play around, but it stops after memory usage creeps up to 1.27GB.
> 
> Any ideas?


okay check environment variable  [here](https://github.com/markshust/docker-magento/blob/master/compose/env/elasticsearch.env)

it should be localhost or 127.0.0.1 for ES_HOST { host name}. it will work or check your ElasticSearch configuration.

**Comment by Asitis** [comment]
That file is left default;
`ES_HOST=elasticsearch
ES_PORT=9200`

If I change the ES_HOST to 127.0.0.1, same result; it returns `Failed to connect to 127.0.0.1 port 9200: Connection refused`

**Comment by markshust** [comment]
Yea, it should *not* be `127.0.0.1`, because that would tell the phpfpm container to look at _itself_ for the Elasticsearch service. The `elasticsearch` hostname tells it to look at the `elasticsearch` container.

@Asitis can you confirm the Elasticsearch container is connectable from the phpfpm container?

You can do that with: `bin/cli curl elasticsearch:9200`

That should output something similar to:

```
{
  "name" : "7e40e6663c15",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "XibYLKClSJ2OhvKW2s_dzw",
  "version" : {
    "number" : "7.9.3",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "c4138e51121ef06a6404866cddc601906fe5c868",
    "build_date" : "2020-10-16T13:34:25.304557Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

I also noticed something odd about your screenshot:

![152498811-202077fa-d799-435a-a72c-8f68231cc43f](https://user-images.githubusercontent.com/437029/152988633-78473d6f-f8e3-475f-83cb-de0a506ab522.png)

That does not look like standard output running `docker ps`, is that from another command, or do you have some sort of third-party Docker thing going on (K8S?)?

My output looks more like:

![Screen Shot 2022-02-08 at 7 39 17 AM](https://user-images.githubusercontent.com/437029/152988813-a9cac4c1-a5e0-4b8e-b2c5-25adfd3a7bb7.png)

It's also possible you are running a command I'm completely ignorant about 😆

**Comment by markshust** [comment]
> > Thats good to hear I'm not the only one, but could you maybe elaborate on your system and setup, things you did and tried? This way hopefully Mark can zero in on the issue.
> 
> i found the answer, just increase docker RAM

Thanks for reminding me about that! This should really, really be detected in the startup or setup script because I've ran into it before. It's an easy mistake to make. I have an open issue about this at https://github.com/markshust/docker-magento/issues/527

**Comment by Asitis** [comment]
> Yea, it should _not_ be `127.0.0.1`, because that would tell the phpfpm container to look at _itself_ for the Elasticsearch service. The `elasticsearch` hostname tells it to look at the `elasticsearch` container.
> 
> @Asitis can you confirm the Elasticsearch container is connectable from the phpfpm container?
> 
> You can do that with: `bin/cli curl elasticsearch:9200`
Nope, this returns the **Connection Refused** it silently puts out during bin/setup:
```
bin/cli curl elasticsearch:9200
curl: (7) Failed to connect to elasticsearch port 9200: Connection refused
```

> That does not look like standard output running `docker ps`, is that from another command, or do you have some sort of third-party Docker thing going on (K8S?)?
To be honest, I'm not sure. I ran `docker ps` again and get an output more like yours

![image](https://user-images.githubusercontent.com/9464641/152993888-b911ec0b-202a-4d91-bd29-da7914a1bc38.png)

**Comment by DasunKanishka** [comment]
I've been using this docker setup for couple of projects in the past. It didn't give much trouble when using on Ubuntu. But now I'm getting _Waiting for connection to Elasticsearch..._ and it breaks the installation. First I tried the Automated Setup and then tried it doing Manually. While running `bin/setup` it gets stuck on Elasticsearch and the breaks the installation.

**My System**
I'm running Ubuntu 20.04, Gnome 3.36.8, 16GB RAM with a i7-8750H CPU

I did do the extra Linux setup. But in the past I didn't have to do any custom docker configurations other than the [post installation steps](https://docs.docker.com/engine/install/linux-postinstall/) to setup this on Linux. The Automated Setup worked nicely.

**Docker ps**
![Screenshot from 2022-02-10 09-04-50](https://user-images.githubusercontent.com/12710251/153332588-a5e99146-6fd5-4320-8c3e-140222ff81c6.jpg)

**bin/cli curl elasticsearch:9200**
`curl: (7) Failed to connect to elasticsearch port 9200: Connection timed out`

**Comment by Asitis** [comment]
> For those with a "failed to connect" message, does this line work?

$ bin/cli curl 172.17.0.1:9200
curl: (7) Failed to connect to 172.17.0.1 port 9200: Connection refused





> I think this is all related to how DNS is (not) working on Linux.
> 
> This may be possibly fixed with this ticket: #636

I'd love to try this, but, I'm not sure how or where? There is no mention of buster nor bullseye in any file or folder?

**Comment by markshust** [comment]
@Asitis if that `curl` line is failing you, that other fix won't help you.

Does that line work for you when running it on your host machine?

`curl localhost:9200`

**Comment by Asitis** [comment]
If I run that command `curl localhost:9200` on my machine, I do get a result;
```
curl localhost:9200
{
  "name" : "eec67631afe6",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "FMp3oNTcTeCinu9-uzyblA",
  "version" : {
    "number" : "7.9.3",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "c4138e51121ef06a6404866cddc601906fe5c868",
    "build_date" : "2020-10-16T10:36:16.141335Z",
    "build_snapshot" : false,
    "lucene_version" : "8.6.2",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

To add to this, not sure if it is relevant; It appears I have a local redis-server installed on my machine, I have to run `sudo service redis-server stop` before running the `bin/setup magento2.test`. If I don't, the output is this:

```docker$ bin/setup magento2.test
Starting docker_elasticsearch_1 ... 
Starting docker_rabbitmq_1      ... 
Starting docker_app_1           ... 
Starting docker_mailcatcher_1   ... 
Starting docker_db_1            ... 
Starting docker_redis_1         ... 
Starting docker_redis_1         ... error
Starting docker_elasticsearch_1 ... done
Starting docker_rabbitmq_1      ... done
Starting docker_app_1           ... done
Starting docker_mailcatcher_1   ... done
Starting docker_db_1            ... done
Error starting userland proxy: error while calling PortManager.AddStarting docker_phpfpm_1        ... done

ERROR: for redis  Cannot start service redis: driver failed programming external connectivity on endpoint docker_redis_1 (840d03f383b2258528cd891f7f1b4d8bd2cbb381687ca73899106ded92dbcd33): Error starting userland proxy: error while calling PortManager.AddPort(): listen tcp4 0.0.0.0:6379: bind: address already in use
ERROR: Encountered errors while bringing up the project.
```

**Comment by markshust** [comment]
I don't understand how `bin/cli curl 172.17.0.1:9200` is failing on a clean start of this project, as that is the IP of the Docker machine and will route to the Elasticsearch service within the container.

You definitely need to have any host services disabled from running on ports 80, 443, 3306, ...9200, and any other service used in this project, as the ports bind-mount back to the host at the time of start.

I'm unable to replicate this, so going to mark as "unable to replicate" unless someone else can confirm they are having the same issue. From what I know, this is running on a clean install of Linux as long as no services are running on conflicting local host ports.

**Comment by Asitis** [comment]
Well OK, thanks for your time anyway. I'll guess I'll try to figure it out or look for a different solution or, worst case scenario, do a clean install of my machine to get it to work..

**Comment by markshust** [comment]
Appreciate it. I have a spare machine finally that I can install Win/Lin on, so I will also test the install myself on a completely fresh install of Ubuntu and report back.

**Comment by Asitis** [comment]
I'll try to debug this as far as I can and report back also. I'd hate to have to do a reinstall, but your Docker setup had become a cornerstone of my workflow so it might be worth it. Anyway, appreciate all your hard work on this, making it possible it the first place! :pray:

**Comment by Asitis** [comment]
FYI: I used the` composer.yml` from a colleague who was using an older version, to downgrade and see what would happen. This time Elasticsearch did pass, but now RabbitMQ started spewing errors all over my terminal. 
I decided to do a clean install of my machine, just set it up again with the latest version without a hitch. I'm guessing it has something to do with me having updated from an earlier version, and then trying to reinstall.
Not a solution perse, but I'm back on the road again. Thanks again, and sorry for everyone else running into the same issues as me and finding this topic!
